### PR TITLE
feat: pnpm import support yarn

### PR DESCRIPTION
**Pull request by fireairforce**
I will continue to finish this pr: https://github.com/pnpm/pnpm/pull/2836

But i have some question:
- Can i transfer `yarn.lock` to a `package-lock.json` file and then to use the logic existed:

```ts
const versionsByPackageNames = {}
  getAllVersionsByPackageNames(npmPackageLock, versionsByPackageNames)
  const preferredVersions = getPreferredVersions(versionsByPackageNames)
  const store = await createOrConnectStoreController(opts)
  const installOpts = {
    ...opts,
    lockfileOnly: true,
    preferredVersions,
    storeController: store.ctrl,
    storeDir: store.dir,
  }
  await install(await readProjectManifestOnly(opts.dir), installOpts)
```

I just need to transfer `yarn.lock` to the `npmPackageLock`.

@zkochan cc~

close #2836

### CODE REVIEW

**Comment by zkochan** [comment]
I don't have objections. If it is easier to convert yarn.lock to package.json then to directly convert yarn.lock to pnpm-lock.yaml

**Comment by wre232114** [comment]
@fireairforce I think we need to support yarn workspaces to pnpm workspace, so I think transform yarn.lock to pnpm lock first 
 is not practical, yarn.lock to preferedVersionMap directly is a more ideal way

**Comment by fireairforce** [comment]
I have finished the development of `pnpm import yarn`, Maybe you can review for me when you free~ @zkochan

File: packages/plugin-commands-import/src/import.ts [hunk 1]
```diff
@@ -49,6 +71,20 @@ export async function handler (
   await install(await readProjectManifestOnly(opts.dir), installOpts)
```
Comment (wre232114) [inline packages/plugin-commands-import/src/import.ts:65]: install only accept one workspace, we may use mutateModules to support yarn workspace to pnpm workspace

Comment (fireairforce) [reply packages/plugin-commands-import/src/import.ts:65]: > install only accept one workspace, we may use mutateModules to support yarn workspace to pnpm workspace

Ok, i will think some way to complete the yarn workspace to pnpm workspace transform, maybe i need a judge here.

Comment (fireairforce) [reply packages/plugin-commands-import/src/import.ts:65]: honestly speaking, can u tell some differences between one workspace and more worksapces lock file?
It seems that the yarn will hoist some deps on the top but pnpm not? 
So i just need to deal with deps of the top `package.json` and then to generate the `pnpm-lock.yaml` file?

File: packages/plugin-commands-import/src/import.ts [hunk 2]
```diff
@@ -96,6 +132,21 @@ function getAllVersionsByPackageNames (
   }
 }
 
+function getAllVersionsByYarnLockFile (
```
Comment (zkochan) [inline packages/plugin-commands-import/src/import.ts]: Probably it should be 
```suggestion
function getAllVersionsFromYarnLockFile (
```

File: packages/plugin-commands-import/test/index.ts [hunk 3]
```diff
@@ -100,3 +128,22 @@ test('import fails when no npm lockfiles are found', async () => {
...
+    await importCommand.handler({
+      ...DEFAULT_OPTS,
+      cliOptions: {
+        yarn: true,
+      },
+      dir: process.cwd(),
+    })
+  } catch (_err) {
+    err = _err
+  }
```
Comment (zkochan) [inline packages/plugin-commands-import/test/index.ts]:
```suggestion
  await expect(
    importCommand.handler({
      ...DEFAULT_OPTS,
      cliOptions: {
        yarn: true,
      },
      dir: process.cwd(),
    })
  ).rejects.toThrow()
```

File: packages/pnpm/package.json [hunk 4]
```diff
@@ -150,7 +150,7 @@
     "bundle:pnpm": "cross-var esbuild lib/pnpm.js --bundle --platform=node --outfile=dist/pnpm.cjs --external:node-gyp --define:process.env.npm_package_name=\\\"$npm_package_name\\\" --define:process.env.npm_package_version=\\\"$npm_package_version\\\"",
     "bundle:pnpx": "esbuild lib/pnpx.js --bundle --platform=node --outfile=dist/pnpx.cjs",
     "bundle": "pnpm bundle:pnpm && pnpm bundle:pnpx",
-    "start": "pnpm tsc -- --watch",
+    "start": "pnpm tsc --watch",
```
Comment (zkochan) [inline packages/pnpm/package.json:153]: why is this needed?

**Review by zkochan — COMMENTED** [review]
Please add changesets that describe the chagnges: `pnpm -w changeset`

File: packages/plugin-commands-import/src/import.ts (hunk 1 above)
Comment (zkochan) [reply packages/plugin-commands-import/src/import.ts:65]: I think it is fine as it is now.

File: packages/plugin-commands-import/src/import.ts [hunk 5]
```diff
@@ -11,33 +12,56 @@ import { install, InstallOptions } from 'supi'
...
+          {
+            description: `Generate ${WANTED_LOCKFILE} from yarn.lock file`,
+            name: '--yarn',
+          },
+        ],
+      },
+    ],
+    usages: [
+      'pnpm import',
+      'pnpm import yarn',
```
Comment (zkochan) [inline packages/plugin-commands-import/src/import.ts]: Is the option needed at all? pnpm can automatically detect whether there is a Yarn or a npm lockfile.

```suggestion
      'pnpm import --yarn',
```

File: packages/pnpm/package.json (hunk 4 above)
Comment (fireairforce) [reply packages/pnpm/package.json:153]: because it doesn't work when add `--`, so i removed it ...

File: packages/plugin-commands-import/src/import.ts (hunk 5 above)
Comment (fireairforce) [reply packages/plugin-commands-import/src/import.ts]: ok

**Comment by fireairforce** [comment]
Updated~PTAL @zkochan

**Comment by fireairforce** [comment]
The pr has been updated and the changeset file added. PTAL @zkochan

File: packages/plugin-commands-import/src/import.ts [hunk 6]
```diff
@@ -11,33 +12,56 @@ import { install, InstallOptions } from 'supi'
...
 ) {
   // Removing existing pnpm lockfile
   // it should not influence the new one
   await rimraf(path.join(opts.dir, WANTED_LOCKFILE))
-  const npmPackageLock = await readNpmLockfile(opts.dir)
   const versionsByPackageNames = {}
-  getAllVersionsByPackageNames(npmPackageLock, versionsByPackageNames)
-  const preferredVersions = getPreferredVersions(versionsByPackageNames)
+  let preferredVersions = {}
+  if (opts?.cliOptions && opts.cliOptions['yarn']) {
```
Comment (zkochan) [inline packages/plugin-commands-import/src/import.ts]: why is this option needed? We could just check if a `yarn.lock` file exists. If it exists, import it, if not, import the npm lockfile.

Comment (fireairforce) [reply packages/plugin-commands-import/src/import.ts]: In Some project, the package-lock.json and yarn.lock file both exist,i just want to make a separate of them.

Comment (zkochan) [reply packages/plugin-commands-import/src/import.ts]: It sounds like an edge case. In those rare cases the user may rename or remove one of the lockfiles before running `pnpm import`

Comment (fireairforce) [reply packages/plugin-commands-import/src/import.ts]: Maybe i think it's normal in the development，Some people even don't konw what the lock file was doing...So i think this option maybe necessary...

Comment (zkochan) [reply packages/plugin-commands-import/src/import.ts]: I don't get it. How is having two lockfiles at the same time is normal? You either use Yarn or you use npm, combining two different package managers is a really bad idea.

Comment (fireairforce) [reply packages/plugin-commands-import/src/import.ts:57]: I mean that if a project was developed by many people...

Comment (fireairforce) [reply packages/plugin-commands-import/src/import.ts:57]: I think it's much better to add a option here, maybe next time Some new option will be added，and users can understand the usage of command more clear.🤔

Comment (fireairforce) [reply packages/plugin-commands-import/src/import.ts]: emm, maybe you are right, i will adjust the logic here~

**Comment by fireairforce** [comment]
The Code has been adjusted, maybe you can review it again @zkochan

**Review by zkochan — APPROVED** [review]

**Comment by zkochan** [comment]
Some tests are failing

**Comment by fireairforce** [comment]
> Some tests are failing

I am trying to fix this, the new commit has been push~
